File: backend/app/schemas/artifact.py
import uuid
from datetime import datetime

from pydantic import ConfigDict, Field

from app.domain.enums import ArtifactType, ParserExecutionStatus
from app.schemas.case import PaginationMeta
from app.schemas.common import ReadSchema


class ParseRequest(ReadSchema):
    model_config = ConfigDict(extra="forbid")

    artifact_type: ArtifactType


class ArtifactRead(ReadSchema):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: uuid.UUID
    evidence_id: uuid.UUID
    artifact_type: ArtifactType
    parser_name: str
    parser_version: str
    status: ParserExecutionStatus
    metadata: dict[str, object] = Field(validation_alias="metadata_")
    warnings: list[str]
    errors: list[str]
    statistics: dict[str, int | float]
    record_count: int
    created_at: datetime


class ArtifactListResponse(ReadSchema):
    data: list[ArtifactRead]
    meta: PaginationMeta


class ArtifactRecordRead(ReadSchema):
    id: uuid.UUID
    artifact_id: uuid.UUID
    record_type: str
    source_record_identifier: str | None
    event_time: datetime | None
    data: dict[str, object]
    provenance: dict[str, object]
    created_at: datetime


class ArtifactRecordListResponse(ReadSchema):
    data: list[ArtifactRecordRead]
    meta: PaginationMeta

File: frontend/src/types/artifact.ts
import type { PaginationMeta } from './case'

export type ProductionArtifactType = 'EVTX' | 'REGISTRY' | 'PREFETCH' | 'LNK' | 'NTFS_MFT'
export type ParserStatus = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'COMPLETED_WITH_WARNINGS' | 'FAILED' | 'UNSUPPORTED'

export interface ArtifactRecord {
  id: string
  evidence_id: string
  artifact_type: ProductionArtifactType | string
  parser_name: string
  parser_version: string
  status: ParserStatus
  metadata: Record<string, unknown>
  warnings: string[]
  errors: string[]
  statistics: Record<string, number>
  record_count: number
  created_at: string
}

export interface ArtifactListResponse { data: ArtifactRecord[]; meta: PaginationMeta }

export interface DerivedArtifactRecord {
  id: string
  artifact_id: string
  record_type: string
  source_record_identifier: string | null
  event_time: string | null
  data: Record<string, unknown>
  provenance: Record<string, unknown>
  created_at: string
}

export interface ArtifactRecordListResponse { data: DerivedArtifactRecord[]; meta: PaginationMeta }
export interface ParseEvidenceRequest { artifact_type: ProductionArtifactType }
